text/1000-migrations.rst

```diff
@@ -76,6 +80,71 @@ whether there are untracked changes in the local schema that require a
 new migration.
 
 
+Migration History
+=================
+
+The migration history is defined as a linear sequence of migrations (although
+this might change in the future to support merging migration histories).
+Each migration has a unique name that is a hash derived from the token
+stream of the migration script and the migration parent name.  To
+make migration names valid identifiers the hash is prefixed with letter
+``'m'``.
+
+The names of the migrations are computed by the ``edgedb migrate`` command
+starting from the latest committed migration (if any), using the following
+algorithm:
+
+1. Get the latest migration id from the database.
+2. Scans all files in the migrations dir and finds a file that references
+   that parent.
+3. If there's no such file --- terminate: there's no in-progress migration.
+4. if there's one such file and it has the largest sequential file name --
+   then there's only one in-progress migration file, so it is committed.
+5. If there's a file with that parent and other files after it it means
+   that we have a few in-progress (not yet committed) migration files.
+   The parent names in them are fluid -- generated and present in the file,
+   but aren't yet in the database.
+   Scenarios:
+   a) A user has two in-progress migration files: 0004.edgeql and 0005.edgeql.
+      If the user modifies 0004.edgeql and runs edgedb migration commit, we
+      compute the name for 0004 and commit it; we then update the parent name
+      in 0005 and commit it too.
```

Comment:
This shouldn't silently update parent commit and apply. We need a flag to do that ``edgedb migrate --allow-rebasing``. And should fail by default.

The reason is this scenario:
1. I've edited file and did not run ``edgedb migrate`` on my laptop
2. Committed my edited file to git
3. Then CI runs ``edgedb migrate`` on production database, silently modifies the migration files and they are never committed from CI
4. Now revisions in git and production are out of sync.

In this scenario, hopefully you can run migrate locally again and get the same results. But you already don't know which ones are on production and whether there were any extra edits after that.

To fix that we should fail at step 3 and force developers to run `--allow-rebase` on their machine.

Reply:
> And should fail by default.

Should the entire operation fail, i.e. both 0004 and 0005 not committed?

Reply:
Yes, if user requested to upgrade to the "head" (last revision). I.e. if we know in advance that we can't fulfill the whole operation there is no reason to partially apply.

If we have command `edgedb migrate --to=0004.sql` then operation should succeed (and *may* give a hint that next revision has different parent).